py/htm/examples/mnist.py

```diff
@@ -15,65 +15,51 @@
-    def int32(b):
-        i = 0
-        for char in b:
-            i *= 256
-            # i += ord(char)    # python2
-            i += char
-        return i
-
-    def load_labels(file_name):
-        with open(file_name, 'rb') as f:
-            raw = f.read()
-            assert(int32(raw[0:4]) == 2049)  # Magic number
-            labels = []
-            for char in raw[8:]:
-                # labels.append(ord(char))      # python2
-                labels.append(char)
-        return labels
-
-    def load_images(file_name):
-        with open(file_name, 'rb') as f:
-            raw = f.read()
-            assert(int32(raw[0:4]) == 2051)    # Magic number
-            num_imgs   = int32(raw[4:8])
-            rows       = int32(raw[8:12])
-            cols       = int32(raw[12:16])
-            assert(rows == 28)
-            assert(cols == 28)
-            img_size   = rows*cols
-            data_start = 4*4
-            imgs = []
-            for img_index in range(num_imgs):
-                vec = raw[data_start + img_index*img_size : data_start + (img_index+1)*img_size]
-                # vec = [ord(c) for c in vec]   # python2
-                vec = list(vec)
-                vec = np.array(vec, dtype=np.uint8)
-                vec = np.reshape(vec, (rows, cols))
-                imgs.append(vec)
-            assert(len(raw) == data_start + img_size * num_imgs)   # All data should be used.
-        return imgs
-
-    train_labels = load_labels(os.path.join(path, 'train-labels-idx1-ubyte'))
-    train_images = load_images(os.path.join(path, 'train-images-idx3-ubyte'))
-    test_labels  = load_labels(os.path.join(path, 't10k-labels-idx1-ubyte'))
-    test_images  = load_images(os.path.join(path, 't10k-images-idx3-ubyte'))
+def load_ds(name, num_test, shape=None):
+    """ 
+    fetch dataset from openML.org and split to train/test
+    @param name - ID on openML (eg. 'mnist_784')
+    @param num_test - num. samples to take as test
+    @param shape - new reshape of a single data point (ie data['data'][0]) as a list. Eg. [28,28] for MNIST
+    """
+    data = fetch_openml(name, version=1)
+    sz=data['target'].shape[0]
+
+    X = data['data']
+    if shape is not None:
+        new_shape = shape.insert(0, sz)
+        X = np.reshape(X, shape)
+
+    y = data['target'].astype(np.int32)
+    # split to train/test data
+    train_labels = y[:sz-num_test]
+    train_images = X[:sz-num_test]
+    test_labels  = y[sz-num_test:]
+    test_images  = X[sz-num_test:]
 
     return train_labels, train_images, test_labels, test_images
 
+def encode(data, out):
+    """
+    encode the (image) data
+    @param data - raw data
+    @param out  - return SDR with encoded data
+    """
+    out.dense = data >= np.mean(data) # convert greyscale image to binary B/W.
+    #TODO improve. have a look in htm.vision etc. For MNIST this is ok, for fashionMNIST in already loses too much information
```

Comment:
OT: but next step in this domain would be looking at better encoders for vision. The current B/W encoding is likely losing too much information

Reply:
Thanks for doing this PR Breznak

On the topic of vision: I wrote a eyeball encoder a while ago, and I haven't had the time to incorporate it into this repo.  Its written in python, using numpy and open-cv.  I will see if I can get some examples of what it does and post them on the encoder topic thread.

Reply:
>  wrote a eyeball encoder a while ago

wow! How does it work in principle? I was thinking about some saccadic/sparse encoder for vision. Please do keep us possted, once you have time. 

Btw, I was looking for what could be salvaged from nupic.vision repo... and nothing much, there's some image encoder with gabon filter...